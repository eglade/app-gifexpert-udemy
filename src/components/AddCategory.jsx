import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({onAddCategory}) => {
    
    // useState input agregar gif
    const [inputValue, setInputValue] = useState('');
    
    /**
     * Cambio al teclear en input principal gif
     * @param {*} e 
     */
    const onInputChange = (e) => {
        setInputValue(e.target.value);
    }

    /**
     * Metodo submit
     * @param {*} e 
     */
    const onInputSubmit = (e) => {
        // no refresh al enviar
        e.preventDefault();

        // validar vacio
        if (inputValue.trim().length <= 0) return;

        // utiliza funcion del padre para eliminas
        onAddCategory(inputValue);
        
        // seteo input a vacio y todo salio bien
        setInputValue('');            
    }    

    /**
     * Retorno componente formulario búsqueda
     */
    return (
        <form aria-label="form" onSubmit={(e) => onInputSubmit(e) }>
            <input 
                type="text" 
                placeholder="Búsqueda de gif" 
                value={inputValue} 
                onChange={(e) => onInputChange(e)} 
            />
            <button type="submit">+ Buscar</button>
        </form>
    )
};

/**
 * VALIDACIONES
 */
 AddCategory.propTypes = {
    onAddCategory: PropTypes.func.isRequired
}  
