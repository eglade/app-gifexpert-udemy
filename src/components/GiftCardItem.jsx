import PropTypes from "prop-types";

export const GiftCardItem = ({title, url}) => {
    return (
        <div aria-label="dfgh" className="card">
            <p>{title}</p>
            <img src={url} alt={title} />
        </div>        
    ) 
}

/**
 * VALIDACIONES
 */
GiftCardItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}   