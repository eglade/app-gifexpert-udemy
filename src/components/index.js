export * from './AddCategory';
export * from './GiftCardItem';
export * from './GridCategory';
export * from './ListCategory';