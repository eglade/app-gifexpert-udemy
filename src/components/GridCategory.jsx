import { useFetchGifs } from "../hook/useFetchGifs";
import { GiftCardItem } from "../components";
import PropTypes from "prop-types";

/**
 * Muestra el Grid de una determinada categoria.
 * @param {*} param0 
 * @returns 
 */
export const GridCategory = ({category, index, onDeleteCategory }) => {

    // hook custom
    const {images, isLoading} = useFetchGifs(category);
    
    /**
     * Llamado a eliminar una categoría (Metodo principal esta en GifExpertApp)
     * @param {*} e 
     * @param {*} index 
     */
     const deleteCategory = (e, index) => {
        // no refresh al enviar
        e.preventDefault();

        // eliminar (metodo del padre)
        if (confirm("Eliminar?") == true) {
            onDeleteCategory(index);
        }
    }

    /**
     * Retorno bloque grid de la categoría
     */
    return (
        <>
            <h2>
                <button onClick={(e) => deleteCategory(e, index)}>x</button> {category}
            </h2>

            { isLoading && (<h2>Cargando...</h2>)}

            <div className="card-grid">
                {
                    images.map( (gif) => (
                        <GiftCardItem 
                            key={gif.id} 
                            title={gif.title}
                            url={gif.url}
                        />                    
                    ))                
                }
            </div>                
        </>
    )
}; 


/**
 * VALIDACIONES
 */
 GridCategory.propTypes = {
    category: PropTypes.string.isRequired
}  



