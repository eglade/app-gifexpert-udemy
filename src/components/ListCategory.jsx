import { GridCategory } from "../components";
/**
 * Listado de todas las categorias
 * @param {*} param0 
 * @returns 
 */
export const ListCategory = ({categories, onDeleteCategory}) => {
    // retorno de categorias
    return (
        <>
            {
                // Recorrido de categorias almacenadas
                categories.map( (category, index) => (

                    // bloque grid categor√≠a
                    <GridCategory
                        key={category}    
                        category={category}
                        index={index} 
                        onDeleteCategory={onDeleteCategory}
                    />
                ))
            }
        </>
    )
}