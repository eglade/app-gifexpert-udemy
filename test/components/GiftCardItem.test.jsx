import { render, screen } from "@testing-library/react";
import { GiftCardItem } from "../../src/components/GiftCardItem";

describe('<GiftCardItem.test />', () => {
    
    // gif imagen de prueba
    const gif = {
        id:'123456',
        title:'Goku',
        url: 'https://media2.giphy.com/media/WOb8EeFziTQNE02WXs/giphy.gif?cid=b6da54b3lwd0fa7q7eoiq5u6w0yyozvrps3c64nbyyu9dosb&rid=giphy.gif&ct=g'
    }
    
    test('El componente debe hacer match con el snapshot', () => {
        const {container} = render(<GiftCardItem 
            key={gif.id} 
            title={gif.title}
            url={gif.url}
        />);
        expect(container).toMatchSnapshot();
    });

    test('El componente debe tener titulo en parrafo <p/>', () => {
        const {container} = render(<GiftCardItem 
            key={gif.id} 
            title={gif.title}
            url={gif.url}
        />);
        expect(container.querySelector('p').innerHTML).toContain(gif.title);
    }); 
    
    test('El componente debe tener imagen y titulo en <img src="..." alt"..." /> ', () => {
        const {container} = render(<GiftCardItem 
            key={gif.id} 
            title={gif.title}
            url={gif.url}
        />);
        
        expect(container.querySelector('img').getAttribute("src")).toContain(gif.url);
        expect(container.querySelector('img').getAttribute("alt")).toContain(gif.title);
    });

    test('El componente debe tener imagen y titulo en <img src="..." alt"..." /> 2', () => {
        render(<GiftCardItem 
            key={gif.id} 
            title={gif.title}
            url={gif.url}
        />);

        const {src, alt} = screen.getByRole('img');

        expect(src).toContain(gif.url);
        expect(alt).toContain(gif.title);
    });    

});