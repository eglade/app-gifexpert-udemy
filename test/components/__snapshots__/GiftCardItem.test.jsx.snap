// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiftCardItem.test /> El componente debe hacer match con el snapshot 1`] = `
<div>
  <div
    aria-label="dfgh"
    class="card"
  >
    <p>
      Goku
    </p>
    <img
      alt="Goku"
      src="https://media2.giphy.com/media/WOb8EeFziTQNE02WXs/giphy.gif?cid=b6da54b3lwd0fa7q7eoiq5u6w0yyozvrps3c64nbyyu9dosb&rid=giphy.gif&ct=g"
    />
  </div>
</div>
`;
